### PR TITLE
Use ProjectItem.Name for file name consistency 

### LINKED ISSUE #472: Open on GitHub sometimes 404s when path contains upper case
- GitHub Extension version: 1.0.18.6
- Visual Studio version: 2015 Update 3

Right click on code windows and select 'GitHub > Open on GitHub'. If the source file or directory contains an upper case char, the GitHub page sometimes opens with a 404.

It appears there is an issue with Visual Studio. If a source file was opened automatically when a solution loads, `Document.Name` will return an all lower case path. If the user explicitly opens the document after the solution loads, the path will contain upper and lower chase chars es expected.

You can see what's happening here. I've executed the `go` method using TestDriven.Net's `Test With > In-Proc` command:

document name

The highlighted `netcoretestrunnerlibrarytests.cs` was opened with the solution, but `Scratch.cs` was opened (or created) after the solution was opened. You can see `ProjectItem.Name` always returns a path with the expected casing, but with `Document.Name` it varies.

I'm guessing you could fix it by changing the following line:
https://github.com/github/VisualStudio/blob/master/src/GitHub.VisualStudio/Menus/LinkMenuBase.cs#L21

to...

`return repo.GenerateUrl(activeDocument.ProjectItem.Name, activeDocument.StartLine, activeDocument.EndLine);`

I hope that's useful. :smile: 

You can find discussion about it on Twitter here:
https://twitter.com/jcansdale/status/760547405926395904

### PR DESCRIPTION
**Pull request by paladique**
Fixes #472 per suggestions.

### CODE REVIEW

File: src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ public class ActiveDocumentSnapshot : IActiveDocumentSnapshot
         public ActiveDocumentSnapshot([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
         {
             StartLine = EndLine = -1;
-            Name = Services.Dte2?.ActiveDocument?.FullName;
+            Name = Services.Dte2?.ActiveDocument?.ProjectItem.Name;
```
Comment (shana) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: What happens when there's no project loaded? Will `ProjectItem` still exist, or will it be null? This needs to be tested by opening a file on a repo when there's no solution opened.

If `ActiveDocument` is not returning the correct file case and if `ProjectItem` can be null, then perhaps a better way of fixing this is to get the full name of the file from the filesystem to get the correct case.

Comment (paladique) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: I gave this a try by making a new solution without any projects and added a .txt file and `ProjectItem` still exists. VS might be considering a solution a project in the case when no project exists, or still considers it a project if it does.

Comment (shana) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: You need to be in a git repo context without a solution open. You can either create an empty git repo and switch to it via the local git repositories list, or if you have a solution open, close it. Then you open or drag a file into the editor and see what if the property is still there.

Comment (paladique) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: `ProjectItem.ContainingProject` references the project where the file is, and turns out to be null when testing it out. However, debugging `Services.Dte2.Solution.Projects` contains 1 item, which I assume is the solution.

It's possible also to find a file through `Services.Dte2.Solution.FindProjectItem("somefile")` which might mean that any file, whether it be in solution or project, is considered a `ProjectItem`

Comment (paladique) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: When you open a file in VS without a solution open, VS creates a temporary solution file so  `ProjectItem` still exists, but the link that's generated points to the latest commit of the git repo that's currently in the context. Looks like a bug there, it probably should generate a link if the solution isn't open, I'll put that in another issue.

Comment (shana) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: Yeah, I got that behaviour too. We should probably be checking the repo where the file is and not the current repo to determine the link to go to. That's likely a bigger change though :P

Comment (shana) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: Oh wait, you mean the file is in the repo but it generates the wrong url? Sounds like a bug with the path ir something?

Comment (paladique) [inline src/GitHub.VisualStudio/Helpers/ActiveDocumentSnapshot.cs]: The file wasn't in the repo. I just had a git repo context with no solution open, and opened a random file on my PC.

**Comment by paladique** [comment]
@shana so the end result of this was that:

- In all known cases/scenarios tested, ProjectItem is never null, (solutions are treated like projects, a temporary solution file is created in solution explorer when there is no solution file)

- A [bug](https://github.com/github/VisualStudio/issues/530) was discovered where using the open/copy link context menu items can generate URLs when opening a file that may not be part of the current git repo context

**Comment by shana** [comment]
✨ 

I'll 👀 this first thing in the morning with a fresh brain 😄

**Comment by shana** [comment]
@paladique Update the test manifest with test cases that check for this issue.

**Pull request merged**